packages/ember-toucan-core/src/components/form/controls/combobox.ts

```diff
@@ -0,0 +1,537 @@
+  OPTION extends Option
+> {
+  Args: {
+    /**
+     * A CSS class to add to this component's content container.
+     * Commonly used to specify a `z-index`.
+     */
+    contentClass?: string;
+
+    /**
+     * Sets the input to an errored-state via styling.
+     */
+    hasError?: boolean;
+
+    /**
+     * Sets the `disabled` attribute of the input.
+     */
+    isDisabled?: boolean;
+
+    /**
+     * Sets the `readonly` attribute of the input.
+     */
+    isReadOnly?: boolean;
+
+    /**
+     * A string to display when there are no results after the user's filter.
+     */
+    noResultsText?: string;
+
+    /**
+     * Called when the user makes a selection.
+     * It is called with the selected option (derived from `@options`) as its only argument.
+     */
+    onChange?: (option: unknown) => void;
+
+    /**
+     * The function called when a user types into the combobox textbox, typically used to write custom filtering logic.
+     */
+    onFilter?: (input: string) => OPTION[];
+
+    /**
+     * `@options` forms the content of this component.
+     *
+     * To support a variety of data shapes, `@options` is typed as `unknown[]` and treated as though it were opaque.
+     * `@options` is simply iterated over then passed back to you as a block parameter (`select.option`).
+     */
+    options?: OPTION[];
+
+    /**
+     * When `@options` is an array of objects, `@selected` is also an object.
+     * The `@optionKey` is used to determine which key of the object should
+     * be used for both filtering and displayed the selected value in the
+     * textbox.
+     */
+    optionKey?: string;
+
+    /**
+     * The currently selected option.  If `@options` is an array of strings, provide a string.  If `@options` is an array of objects, pass the entire object.
+     */
+    selected?: OPTION;
+  };
+  Blocks: {
+    default: [
+      {
+        option: OPTION;
+        Option: WithBoundArgs<
+          typeof OptionComponent,
+          | 'index'
+          | 'isActive'
+          | 'isDisabled'
+          | 'isReadOnly'
+          | 'onClick'
+          | 'onMouseover'
+          | 'popoverId'
+        >;
+      }
+    ];
+  };
+  Element: HTMLInputElement;
+}
+
+export default class ToucanFormComboboxControlComponent<
+  OPTION extends Option
+> extends Component<ToucanFormComboboxControlComponentSignature<OPTION>> {
+  @tracked activeIndex: number | null = null;
+  @tracked inputValue: string | undefined;
+  @tracked isPopoverOpen = false;
+  @tracked filteredOptions: OPTION[] | undefined;
+
+  Chevron = Chevron;
+  Option = OptionComponent;
+  popoverId = `popover--${guidFor(this)}`;
+
+  constructor(
+    owner: unknown,
+    args: ToucanFormComboboxControlComponentSignature<OPTION>['Args']
+  ) {
+    super(owner, args);
+
+    // We need to set our input tag's value attribute
+    // if we have a selected option provided on render
+    let { selected, optionKey } = this.args;
+
+    this.inputValue =
```

Comment:
I was expecting `inputValue` to be derived state. Do we really need to handle that state imperatively? Basically the logic below, but as a getter...

Reply:
Let me give this a shot!

Reply:
I'm going to keep playing with this in a downstream PR if that's okay so that the changes can be focused rather than keep adding onto the pile here